### ISSUE TITLE
Support for 3D mice (3D Connection Space Navigator)

### ISSUE THREAD

**Issue by mankoff** [CloudCompare/CloudCompare#6]

**Comment by sendnic** [comment]
It has been some time from last reply, is there a way to get the space mouse working on OSX?

**Comment by cloudcompare** [comment]
Two solutions:
- either someone has a Mac OS, a 3D mouse and can integrate the SDK in the Mac version
- or someone sends a 3D mouse to Andy Maloney :+1:

**Comment by sendnic** [comment]
I just started looking into CloudCompare and it looks promising. Looking for software that could be used with a scanner I am developing http://www.laserscanningforum.com/forum/viewtopic.php?f=69&t=10065&p=53323#p53323 . I will be doing some testing on following weeks and if all goes well sending a 3D mouse if that all it takes could be well possible option.

**Comment by cloudcompare** [comment]
The new version 2.7 is about to be release. Therefore the 3D mouse won't be supported in the Mac OS version but you'll see that it has evolved a lot (better OpenGL support, etc.).

**Comment by sendnic** [comment]
That is a good new. I think I need to take my old windows machine also in use... Does it run on windows 7?

**Comment by cloudcompare** [comment]
Yes it works on all version of Windows (but Vista maybe)

**Comment by sendnic** [comment]
This is off topic(3D mouse) but what do you think about merging point cloud and panorama as proposed here http://www.laserscanningforum.com/forum/viewtopic.php?f=69&t=10065&p=53336#p53336 , is that something cloud compare already supports?

**Comment by dgirardeau** [comment]
Nope! But as it's an open source project, any can contribute of course ;)

2016-03-07 8:45 GMT+01:00 sendnic <notifications@github.com>:

**Comment by sendnic** [comment]
So interested SW developers are called. I could maybe provide few scanners for this purpose during the spring. If you are interested, please contact me via www.levitezer.com (contact)

**Comment by Patola** [comment]
Can you implement this support for Linux too? I have a space navigator and I do not use other OSes. Blender and FreeCAD both use it and have nice configuration dialogs.

**Comment by ruffsl** [comment]
I agree with @Patola .
Should this issue be reopened to address Linux support, we should we open a new ticket?

**Comment by rjanvier** [comment]
@ruffsl do you want to work on this?

**Comment by ruffsl** [comment]
From what I understand from glancing at a82872cdaa4763b03f3475523573bd0cd0642d1a and 9f4db03f1f5b99c5abef78262d5b1c52f86887b2  , cloud compare now uses the official 3DConnexion SDK, yes?

In our group, we have the 3DxWare driver for [`Linux (x86-64) 1.8.0 06/26/2014`](https://www.3dconnexion.com/service/drivers.html) working with 3DConnexion's supplied demos on Ubuntu 16.04. However when checking cloudcompare, the mouse does not seem to be listed in the menu, as [in the wiki](http://www.cloudcompare.org/doc/wiki/index.php?title=3D_mouse_-_Enable):

> This tool is accessible via the 'File > 3D mouse > Enable' 3dconnexion logo.jpg menu.

Are the linux snaps not compiled with the 3DConnexion SDK enabled? I'm not sure of the state of 3DConnexion linux support today, as topics posted to the [official discussion forum](https://www.3dconnexion.com/forum/viewforum.php?f=22) seem active, but the company seems rather reluctant to support anything new:

>Currently there are no plans in the near future to include support for our newest devices in the Linux driver.

Could a free, compatible alternative for 3Dconnexion's 3D input device drivers and SDK, like [spacenav](http://spacenav.sourceforge.net/) be used for cloud compare on linux? Spacenav seems to work fine for other open source 3D projects like Blender, GazeboSim, FreeCAD...

### System info:

```
$ uname -a
Linux my-pc 4.13.0-36-generic #40~16.04.1-Ubuntu SMP Fri Feb 16 23:25:58 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

$ snap list 
Name          Version    Rev   Tracking  Developer     Notes
cloudcompare  2.10-beta  131   stable    cloudcompare  -
```

**Comment by asmaloney** [comment]
Sorry about closing this issue - it was accidentally closed by my commit comment about a Travis fix.

**Comment by rjanvier** [comment]
@ruffsl Thanks to pointing me to spacenav: I will take a look at it.

**Comment by Patola** [comment]
If you have any code for that I'd be glad to test, I have a basic Space Navigator wired 3d mouse that I use with FreeCAD and Blender.

**Comment by rjanvier** [comment]
Unfortunately, at first look, libspnav and 3DConnexion SDK do not seems to be API compatible... so even if I can integrate libspnav to CC, it will be a pain in the *ss to debug without owning one 3D mouse myself. So it would be nice if any of the present happy 3DConnexion devices owner wanted to contribute.

**Comment by Patola** [comment]
I might try. Do you have any directions on how to do it?

**Comment by rjanvier** [comment]
In CC the navigation is handled by by [this class](https://github.com/CloudCompare/CloudCompare/blob/master/qCC/devices/3dConnexion/Mouse3DInput.cpp). basically you have to intercept X11 Events instead of Windows events  and feed them to libspnav functions. For inspiration : 
In freecad wiring is handled at the application level [see here](
https://github.com/FreeCAD/FreeCAD/blob/57e71e5d55b57f67c611a86f4a0b14d683d86ea0/src/Gui/GuiApplicationNativeEventAware.h)
and navigation is handled [here](https://github.com/FreeCAD/FreeCAD/blob/01f7894b2bf2ce50745e7939833dcdaf746d8c88/src/Gui/Quarter/SpaceNavigatorDevice.cpp)

**Comment by runxel** [comment]
Did someone manage to get a 3dconnexion mouse to Andy yet?
This still seems unresolved for MacOS.

**Comment by dgirardeau** [comment]
I don't think so. And the macOS packages are now created by @tmontaigu (but I don't know if he has the time or the will to do something like that!).

**Comment by tmontaigu** [comment]
Yeah i'll probably never have time to work on that

**Comment by pivot69** [comment]
> Yeah i'll probably never have time to work on that

😭

**Comment by BenTristem** [comment]
I for one would love to see SpaceMouse support

**Comment by somtri** [comment]
3D mouse support is in master. From the CHANGELOG, under v2.14.beta:

> - 3D mouse support on macOS, Linux and Windows (thanks to https://github.com/braunsi23 and Paul Rascle!)
> - on Windows, option to compile with the 3DxWare SDK or the generic hidapi library

@BenTristem that covers the SpaceMouse request above.

Worth closing, unless something specific is still missing.

**Comment by runxel** [comment]
In which beta version has this landed? It's not yet in the latest macOS build.

**Comment by prascle** [comment]
The latest beta version (today, CloudCompare_20260827) works with the 3D mouses.

**Comment by runxel** [comment]
Thanks, but as far as I can see it this issue should not be closed yet?

>  If you have moved the cap and nothing happens, make sure the 3DConnexion driver (3DConnexionHelper.app) is not running - it intercepts the device and prevents the HID path from seeing reports.

This reads to me as just an intermediary step / workaround until proper support is enabled, right?
Other apps are dependend on the driver running, so disabling is a nogo.

**Comment by prascle** [comment]
Hello,
You can use both the 3Dconnexion and Hidapi drivers, but not at the same time. Unless you need to use CloudCompare and another application with the 3D mouse simultaneously, this should still be manageable, with just a few extra commands.
If 3DconnexionHelper is running, you must stop it manually before launching CloudCompare (for example, from the Activity Monitor or from a terminal).
If it’s not running and you need it, you can launch it manually from the Dock.
If you use CloudCompare more often than applications that require the 3Dconnexion driver, you can disable the automatic launch of 3DconnexionHelper at login (in which case you’ll need to launch it manually from the Dock when necessary).

Best regards,
Paul

**Comment by somtri** [comment]
@runxel is right that there's still a gap, and I think it's a separate one from the original request.

The 2012 issue asked for 3D mouse support, and that shipped in 2.14.beta. The SpaceMouse works on macOS now. The message @runxel quoted is CloudCompare's own warning: on macOS the hidapi path can't share the device with the running 3Dconnexion driver, so for now 3DConnexionHelper has to be quit by hand.

Letting the two run together, without that manual step, is a real enhancement, but it's a different piece of work from the support that landed. Might be worth tracking on its own rather than reopening this 2012 thread. That's for @prascle and the maintainers to weigh.

**Comment by runxel** [comment]
Hi Paul, thanks for your answer.

> Unless you need to use CloudCompare and another application with the 3D mouse simultaneously, this should still be manageable

Exactly that is sadly my need... so not really feasible for me.

Just to understand the technical reasoning: Why is CC having issues when the driver is running? Isn't the driver necessary for it to work? Or is this like a hack where the raw data is used by CC? Or to ask differently, why does ever other app need the driver?
And since every other app works just fine, what would it take to make CC work correctly?

**Comment by prascle** [comment]
Without being certain of exactly how it works at the low level, I’ll try to offer my interpretation.

In our case, the driver handles the low-level events generated by the 3D mouse and provides more standard system events that are easier for the code (Cloudcompare) responsible for interpreting them to handle (in this case, in terms of view translation and rotation). 3DConnexion provides a driver (3DConnexionHelper).

If you look at the ticket history, you’ll see that there have been numerous issues related to the various supported platforms (Windows, Linux, macOS), driver updates, and the availability of developers to adapt CloudCompare to the driver on each platform.

The hidapi driver was developed independently of 3DConnexion and offers an alternative solution for connecting the 3D mouse. When @Braunsi23 proposed a hidapi-based solution (pull request #2339), we were very interested in integrating it. Furthermore, it works on all platforms with the same code.

From what I understand, the 3DConnexion driver intercepts the low-level events generated by the 3D mouse. These events are then no longer available to be interpreted by another driver. We have no way to change this behavior.

So, if we want to use CloudCompare and another 3D application that works with the 3DConnexion driver at the same time, the only solution I can think of is to adapt CloudCompare’s code to the 3DConnexion driver’s interface. This should be feasible, but I haven’t yet estimated how long it would take to develop.

If you're interested, I recommend opening a new ticket to explain why the current solution doesn't work for you, including an example of how two codes are used simultaneously—but I'm not sure we'll be able to find a developer available to handle it.

Regards,
Paul

**Comment by braunsi23** [comment]
Hi @runxel — I implemented the hidapi path in #2339, so let me try to answer your technical questions precisely.

**Why does CC have issues when the 3DConnexion driver is running?**

CloudCompare's hidapi path reads the raw HID reports directly from the device. On macOS, HID is normally a broadcast model — multiple processes can open the same device and each receives its own copy of the input reports. However, 3DConnexionHelper.app opens the device with an exclusive device seize (kIOHIDOptionsTypeSeizeDevice). Once that seize is active, other processes can still open the device (which is why CloudCompare reports it found the mouse), but they receive zero reports — the helper is the sole consumer of the raw data. That's exactly the situation our warning message refers to, and it's by design in 3DConnexion's driver, not something we can change from the client side.

**Isn't the driver necessary for it to work? Why does every other app need the driver?**

The other apps you're thinking of (Blender, Fusion 360, etc.) don't read raw HID data at all — they integrate the proprietary 3DConnexion SDK. With that architecture, the helper daemon is required: it transforms the raw reports into higher-level events and forwards them via IPC to whichever application currently has focus. The driver arbitrates between apps instead of fighting over the device. As a bonus, apps get configurable per-application button profiles in the 3DConnexion control panel.

CloudCompare's hidapi path deliberately does the opposite: it consumes the raw HID reports itself, needs no proprietary SDK, no vendor registration, and — importantly — works with the same code on macOS, Linux and Windows. That's what finally closed this 14-year-old issue, since the macOS SDK was discontinued for years and only recently became available again.

**What would it take to make CC work alongside the driver?**

Essentially: implement the 3DConnexion SDK interface as a third backend (alongside 3DxWare on Windows and hidapi everywhere). Then CC would register with the helper like Blender does, and the driver would route events to whichever app is in the foreground — giving you exactly the simultaneous-use workflow you need. It's feasible, but it means acquiring the proprietary SDK framework, a license review for a GPL project, and a new platform-specific code path — which is why we started with hidapi. If that's a workflow you'd depend on, +1 on Paul's suggestion to open a dedicated ticket for it so we can gauge the demand.

Cheers, Steffen
